### PR TITLE
chore: modernize project to latest defaults

### PR DESCRIPTION
**Pull request by AviVahl**
This PR contains a:

- [ ] **bugfix**
- [ ] new **feature**
- [x] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case
Use latest infrastructure `webapck-contrib` has to offer.
Modernize project a bit.

### Breaking Changes
- dropped Node<8.9.0.
- added `peerDependency` of webpack^4.0.0.

### Additional Info
- now uses esm imports/exports.
- converted EnhancedMocha to a class.
- added options schema and validation.

### CODE REVIEW

**Comment by AviVahl** [comment]
@evilebottnawi
hey! I've upgraded the project to latest `webpack-defaults`, but I'm not sure how to activate Azure Pipelines. Best guess it requires admin privileges. Any chance you can assist? :)
 
In addition, and if possible, I'd be happy if you could review this... saw you did that for several other loaders, so you're probably the best person to do that.

Thanks

**Comment by AviVahl** [comment]
I've tested on a Mac (I personally use Linux) and it failed.
I'll refactor the test to not check the modules on the stats... but the actually bundled output.

**Comment by evilebottnawi** [comment]
@AviVahl thanks for help, WIP on azure pipelines

File: README.md [hunk 1]
```diff
@@ -75,46 +75,42 @@ module.exports = {
...
 [npm-url]: https://npmjs.com/package/mocha-loader
-
 [node]: https://img.shields.io/node/v/mocha-loader.svg
 [node-url]: https://nodejs.org
-
 [deps]: https://david-dm.org/webpack-contrib/mocha-loader.svg
 [deps-url]: https://david-dm.org/webpack-contrib/mocha-loader
-
-[test]: 	https://img.shields.io/circleci/project/github/webpack-contrib/mocha-loader.svg
+[test]: https://img.shields.io/circleci/project/github/webpack-contrib/mocha-loader.svg
```
Comment (evilebottnawi) [inline README.md]: Need fix link on badge
```
[![Build Status](https://dev.azure.com/webpack-contrib/mocha-loader/_apis/build/status/webpack-contrib.mocha-loader?branchName=master)](https://dev.azure.com/webpack-contrib/mocha-loader/_build/latest?definitionId=20&branchName=master)
```

File: src/index.js [hunk 2]
```diff
@@ -1,26 +1,32 @@
...
+export default function loader() {}
+
+const startScriptPath = path.join(__dirname, 'start.js');
+const webScriptPath = path.join(__dirname, 'web.js');
+const enhancedMochaPath = path.join(__dirname, 'EnhancedMocha.js');
+
+export function pitch(req) {
+  const options = getOptions(this) || {};
+
+  validateOptions(schema, options, 'mocha-loader');
```
Comment (evilebottnawi) [inline src/index.js]:
```js
validateOptions(schema, options, {
    name: 'Mocha Name',
    baseDataPath: 'options',
  });
```

File: test/__snapshots__/loader.test.js.snap [hunk 3]
```diff
@@ -1,9 +1,19 @@
...
+if(module.hot) {
+	module.hot.accept();
+	module.hot.dispose(function() {
+		mocha.suite.suites.length = 0;
+		var stats = document.getElementById('mocha-stats');
+		var report = document.getElementById('mocha-report');
+		stats && stats.parentNode.removeChild(stats);
+		report && report.parentNode.removeChild(report);
+	});
+}"
```
Comment (evilebottnawi) [inline test/__snapshots__/loader.test.js.snap]: It is expected?

**Comment by evilebottnawi** [comment]
Also need fix CI problem on windows, just create helpers what remove absolute paths

File: test/__snapshots__/loader.test.js.snap (hunk 3 above)
Comment (AviVahl) [reply test/__snapshots__/loader.test.js.snap]: no, it has to do with the new webpack helper. I'll refactor the test to verify output bundle.

**Comment by AviVahl** [comment]
> Also need fix CI problem on windows, just create helpers what remove absolute paths
now that I have CI output, I will fix and ping you back... :)

File: src/index.js [hunk 4]
```diff
@@ -18,7 +18,10 @@ mochaLoader.pitch = pitch;
 export function pitch(req) {
   const options = getOptions(this) || {};
 
-  validateOptions(schema, options, 'mocha-loader');
+  validateOptions(schema, options, {
+    name: 'mocha-loader',
```
Comment (evilebottnawi) [inline src/index.js]: `Mocha Loader`

**Comment by evilebottnawi** [comment]
@AviVahl just ping me when it was ready, thanks for helping :+1:

**Comment by AviVahl** [comment]
@evilebottnawi Thanks! It feels ready for re-review.

I do think the current test is way too naive and proper e2e one is needed. Maybe even one that launches puppeteer/webpack-dev-server and verifies output bundles actually run/work.
I guess that could wait for future PRs though. This one is big enough as it is.

**Comment by AviVahl** [comment]
I've just realized the options schema must include all serializable mocha options. Otherwise it will fail when people try customizing `timeout` and other such fields. I'll update the PR.

**Pull request merged**